### PR TITLE
Add package to provide gevent compatibiliy (#628)

### PR DESCRIPTION
**Pull request by aberres**

### CODE REVIEW

**Comment by reyang** [comment]
Thanks @aberres!

Having this in the core library README is a bit weird (as one can imagine, eventually we'll end up with workarounds for all kinds of 3rd party stuff in the main README, which could confuse/scare the audiences), probably having an extension `opencensus-ext-gevent` makes more sense? @aberres @c24t 

What's in my mind:
1. Having `opencensus-ext-gevent` package.
2. Specify the `gevent` version in `setup.py` (so that in the future if gevent supports async, people won't install the wrong package and flip the context back to thread-local-storage by mistake).
3. In the `__init__.py`, flip the context.
4. Mention in the package README how to use it properly (e.g. needs to be imported before any other OpenCensus components).

**Comment by aberres** [comment]
@reyang fair enough.

I just pushed a commit containing a new package which makes things "just work" when it is installed.
After gevent did its patching matching a setuptools entry point is called. See http://www.gevent.org/api/gevent.monkey.html#plugins for details.

I did not polish this yet, I want to check first of this is more to your liking.

File: contrib/opencensus-ext-gevent/README.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
+OpenCensus gevent Exporter
```
Comment (reyang) [inline contrib/opencensus-ext-gevent/README.rst]: Minor: remove the "Exporter"? Probably this is "OpenCensus gevent Integration"?

File: contrib/opencensus-ext-gevent/README.rst [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+::
+
+    pip install opencensus-ext-gevent
+
+Usage
+-----
+
+The **OpenCensus gevent helper** ...
+
+TBD
```
Comment (reyang) [inline contrib/opencensus-ext-gevent/README.rst]: @aberres This is the good place to put the original doc that you prepared in the README.

File: contrib/opencensus-ext-gevent/setup.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        'Programming Language :: Python :: 3.4',
+        'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: 3.7',
+    ],
+    description='OpenCensus gevent compatibility helper',
+    include_package_data=True,
+    long_description=open('README.rst').read(),
+    install_requires=[
+        'opencensus >= 0.1.dev0, < 1.0.0', 'gevent >= 1.3'
```
Comment (reyang) [inline contrib/opencensus-ext-gevent/setup.py]: Minor, put 'gevent' in a new line.

File: contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+def patch_opencensus(event):
+    # Switch from the default runtime context using ContextVar to one using thread
+    # locals. Needed until gevent supports ContextVar.
+    # See https://github.com/gevent/gevent/issues/1407 and
+    # https://github.com/census-instrumentation/opencensus-python/issues/628
+    import opencensus.common.runtime_context as runtime_context
+
+    runtime_context.RuntimeContext = runtime_context._ThreadLocalRuntimeContext()
+
+    logging.warning("OpenCensus patched for gevent compatibility!!!!!!!")
```
Comment (reyang) [inline contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py]: `logger.warning` or `logging.warning`?

File: README.rst [hunk 5]
```diff
@@ -168,6 +168,28 @@ information, please read the
         }
     }
 
+
```
Comment (reyang) [inline README.rst]: Now we can put the gevent extension link in this section: https://github.com/census-instrumentation/opencensus-python#extensions.

We need to add it in the test script https://github.com/census-instrumentation/opencensus-python/blob/master/nox.py#L33, and have test cases.

**Comment by reyang** [comment]
> I just pushed a commit containing a new package which makes things "just work" when it is installed.
> After gevent did its patching matching a setuptools entry point is called. See http://www.gevent.org/api/gevent.monkey.html#plugins for details.

This approach is amazing!

> I did not polish this yet, I want to check first of this is more to your liking.

This is great, thanks for the contribution @aberres!

File: contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py (hunk 4 above)
Comment (c24t) [reply contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py]: One "!" here is probably sufficient.

File: contrib/opencensus-ext-gevent/setup.py (hunk 3 above)
Comment (c24t) [reply contrib/opencensus-ext-gevent/setup.py]: We changed all the ext dependencies to the latest version of opencensus in https://github.com/census-instrumentation/opencensus-python/pull/630#pullrequestreview-229866858, this should be `0.6.dev0` to be consistent with the others.

File: contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+def patch_opencensus(event):
+    # Switch from the default runtime context using ContextVar to one using thread
+    # locals. Needed until gevent supports ContextVar.
+    # See https://github.com/gevent/gevent/issues/1407 and
+    # https://github.com/census-instrumentation/opencensus-python/issues/628
+    import opencensus.common.runtime_context as runtime_context
```
Comment (jamadden) [inline contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py]: If you wanted to attempt some future-proofing, you could surround this with `if not gevent.monkey.is_module_patched('contextvars'):` 

I'm hoping to get some form of a monkey-patch for contextvars implemented for gevent 1.5.

File: README.rst (hunk 5 above)
Comment (reyang) [reply README.rst]: FYI, I just updated the CONTRIBUTING document:

https://github.com/census-instrumentation/opencensus-python/blob/master/CONTRIBUTING.md#additional-instructions-for-extension-developers

Comment (aberres) [reply README.rst]: Added a test which triggers the gevent patching and checks that the context is switched as expected.

File: nox.py [hunk 7]
```diff
@@ -46,6 +47,7 @@ def _install_dev_packages(session):
 
 
 def _install_test_dependencies(session):
+    session.install('gevent')
```
Comment (reyang) [inline nox.py]: This is not needed. `_install_test_dependencies` is intended for test harness and utils.
`gevent` will be installed by the line you added - `session.install('-e', 'contrib/opencensus-ext-gevent')`.

Comment (aberres) [reply nox.py]: Ah, right. Thanks.

File: contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    if not gevent.monkey.is_module_patched("contextvars"):
+        runtime_context.RuntimeContext = (
+            runtime_context._ThreadLocalRuntimeContext()
+        )
+
+        logging.warning("OpenCensus patched for gevent compatibility")
+    else:
+        logging.warning(
```
Comment (reyang) [inline contrib/opencensus-ext-gevent/opencensus/ext/gevent/geventcompatibility.py]: nice!

**Comment by reyang** [comment]
Thanks @aberres for the great work!

**Comment by allenjohnashton** [comment]
How long until this package hits pypi?

**Comment by reyang** [comment]
@allenjohnashton sorry for the delay, here you go https://pypi.org/project/opencensus-ext-gevent/.

In general we try to have at least a release per month.

**Comment by allenjohnashton** [comment]
Thanks!

**Pull request merged**